### PR TITLE
Qwen2vl vision encoder fix

### PR DESCRIPTION
**Pull request by jakep-allenai**
Potential fix for https://github.com/sgl-project/sglang/issues/2112

## Motivation

Users have reported worse performance running qwen2-vl in sglang and vllm than with transformers. I have identified a few cases of different calculations in the vision encoder. And now, it should be matching perfectly. 

Currently this is a draft PR, because performance is reduced by half roughly.

## Modifications

- Reverted to exactly the same QuickGELU implementation as HF transformers.
- Fixed weird casting issues with attention module used in vision network
- Going back to Torch SDPA

## Checklist

- [ ] Format your code according to the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/references/contributor_guide.md).
- [ ] Add unit tests as outlined in the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/references/contributor_guide.md).
- [ ] Update documentation as needed, including docstrings or example tutorials.

### CODE REVIEW

File: python/sglang/srt/models/qwen2_vl.py [hunk 1]
```diff
@@ -30,10 +30,12 @@
 import torch.nn as nn
 import torch.nn.functional as F
 from einops import rearrange, repeat
+from vllm.config import CacheConfig, MultiModalConfig
```
Comment (merrymercy) [inline python/sglang/srt/models/qwen2_vl.py:33]: remove unused imports

File: python/sglang/srt/models/qwen2_vl.py [hunk 2]
```diff
@@ -30,10 +30,12 @@
...
 from vllm.distributed import parallel_state
 from vllm.distributed import utils as dist_utils
 from vllm.logger import init_logger
 from vllm.model_executor.layers.activation import QuickGELU
+from vllm.model_executor.model_loader.weight_utils import default_weight_loader
```
Comment (merrymercy) [inline python/sglang/srt/models/qwen2_vl.py:38]: remove unused imports

File: python/sglang/srt/models/qwen2_vl.py [hunk 3]
```diff
@@ -201,21 +211,30 @@ def forward(
...
+            attention_mask[
+                ...,
+                cu_seqlens[i - 1] : cu_seqlens[i],
+                cu_seqlens[i - 1] : cu_seqlens[i],
+            ] = True
+
+        q = q.squeeze(0)
+        k = k.squeeze(0)
+        v = v.squeeze(0)
+        output = F.scaled_dot_product_attention(q, k, v, attention_mask, dropout_p=0.0)
```
Comment (merrymercy) [inline python/sglang/srt/models/qwen2_vl.py:234]: I think pytorch SDPA should also be fast, the probably is probably how you prepare the attention mask?
Can you vectorize the code more, use less Python for-loop, or write a triton kernel for it (see [example](https://github.com/sgl-project/sglang/blob/84d96b3ae52ebf65baa6557647e09488b28eee3b/python/sglang/srt/managers/schedule_batch.py#L747-L764)), or catch the results so we can reuse it across layers?

File: python/sglang/srt/models/qwen2_vl.py [hunk 4]
```diff
@@ -52,7 +54,15 @@
...
 
+
 # === Vision Inputs === #
+class OriginalQuickGELUActivation(nn.Module):
+    """
+    Applies GELU approximation that is fast but somewhat inaccurate. See: https://github.com/hendrycks/GELUs
+    """
+
+    def forward(self, input: torch.Tensor) -> torch.Tensor:
+        return input * torch.sigmoid(1.702 * input)
```
Comment (merrymercy) [inline python/sglang/srt/models/qwen2_vl.py:65]: try torch.compile to fuse them?

File: test/srt/test_qwen2vl.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+from sglang.srt.managers.schedule_batch import Req, ScheduleBatch
+from sglang.srt.model_executor.forward_batch_info import ForwardBatch
+from sglang.srt.model_executor.model_runner import ModelRunner
+from sglang.srt.sampling.sampling_params import SamplingParams
+from sglang.srt.server_args import PortArgs, ServerArgs
+
+QWEN2_VL_MODEL = "Qwen/Qwen2-VL-7B-Instruct"
+
+
+class RawSGLangTest(unittest.IsolatedAsyncioTestCase):
```
Comment (merrymercy) [inline test/srt/test_qwen2vl.py:30]: This looks good. Maybe give it a better name. 

Now we have a very good reference script for text-only models and a very good model support guide:
https://github.com/sgl-project/sglang/blob/main/scripts/playground/reference_hf.py
https://sgl-project.github.io/references/supported_models.html#how-to-support-a-new-model

Are you willing to help here to add some scripts/docs similar to the above ones, but for vision language models?

**Review by merrymercy — COMMENTED** [review]
Thanks for investigating this!

File: python/sglang/srt/models/qwen2_vl.py (hunk 3 above)
Comment (jakep-allenai) [reply python/sglang/srt/models/qwen2_vl.py:234]: Hmm, I have tried caching the attention mask, but it doesn't seem to impact performance much.

The issue I see is that for one layer, the `context_attention_fwd` kernel in sglang matches torch's `scaled_dot_product_attention` pretty closely, within 1e-2 for each activation. But, in qwen2vl, there are 32 layers, and after a while, the absolute difference accumulates higher, closer to +/- 1.0 max absolute difference in the activations.

**Comment by merrymercy** [comment]
@jakep-allenai Do you have any updates on this? Qwen2vl is a very popular model so we would like to fix it soon.

**Comment by jakep-allenai** [comment]
No, my implementation with F.fused_dot_product_attention was still 1/2 the speed after caching, and even then, I never heard back from @Mr-Loevan about rerunning his benchmark to see if it would fix his reported issue. On our side, we found no significant difference in user-preference of generations with vllm (which used the xformers backend) or with sglang.

My current theory is that the memory-efficient attention implementation in sglang is accurate enough for a single layer, but small errors will accumulate for a typical 30+ layer full network.

**Comment by yizhang2077** [comment]
I think this pr maybe can be close if #2977 has merged into main since it has fixed encoder. Is it ok for you? @jakep-allenai

**Comment by merrymercy** [comment]
Let us close this one.
